### PR TITLE
Composite checkout: Add postal code auto-formatting

### PR DESCRIPTION
**Pull request by sirbrillig**
#### Changes proposed in this Pull Request

This adds postal code formatting in composite checkout to be similar to the behavior in old checkout.

Old checkout modifies the postal code field value itself when the field is blurred. This PR instead formats only when the postal code value sending to the validation and transaction endpoints. This late formatting may be necessary since there are situations where the field is never edited (eg: initial validation on load to skip the step), or it might be edited before we know the country (which is required for the formatting).

Fixes https://github.com/Automattic/wp-calypso/issues/44622

#### Testing instructions

- Add a domain to your cart and visit checkout.
- Choose The Netherlands as your country.
- Enter the postal code `2651BS` (do not type a space between the numbers and letters).
- Press "Continue".
- Verify that no validation errors are reported for the postal code field.
- Complete the purchase and watch the network request to the transactions endpoint; verify that the `domain_details.postal_code` field contains the postal code `2651 BS`, with a space between the numbers and letters.

### CODE REVIEW

File: client/my-sites/checkout/composite-checkout/composite-checkout.js [hunk 1]
```diff
@@ -397,78 +387,6 @@ export default function CompositeCheckout( {
 				serverAllowedPaymentMethods,
 		  } );
 
-	const renderDomainContactFields = (
```
Comment (michaeldcain) [inline client/my-sites/checkout/composite-checkout/composite-checkout.js:400]: Thanks for moving this to the contact form - it really cleans up this file.

File: client/my-sites/checkout/composite-checkout/payment-method-helpers.js [hunk 2]
```diff
@@ -348,3 +351,9 @@ export function filterAppropriatePaymentMethods( {
 			);
 		} );
 }
+
+export function getPostalCode() {
+	const countryCode = select( 'wpcom' )?.getContactInfo?.()?.countryCode?.value ?? '';
+	const postalCode = select( 'wpcom' )?.getContactInfo?.()?.postalCode?.value;
```
Comment (michaeldcain) [inline client/my-sites/checkout/composite-checkout/payment-method-helpers.js:357]: Does this mean that `postalCode` can be undefined (and `tryToGuessPostalCodeFormat()` expects a string)?

**Comment by sirbrillig** [comment]
Good point. I had originally not applied the formatting to the raw value at all, instead only formatting it before sending it to the server, but I ended up adding it because the function that updates the value as you type (`prepareDomainContactDetails()`) ends up being the same function used in `getDomainDetails()` when preparing the data to be sent to the transactions endpoint. I may be able to split up those two instances instead.

**Comment by sirbrillig** [comment]
I modified the code such that the reformatting occurs only when transmitting the postal code, never for the UI. Does that seem ok? We could also add reformatting on blur, but that seems ultimately kind of unnecessary to me?

**Comment by klimeryk** [comment]
> I modified the code such that the reformatting occurs only when transmitting the postal code, never for the UI. Does that seem ok? We could also add reformatting on blur, but that seems ultimately kind of unnecessary to me? Happy to get disagreement.

IMHO, I think on blur (and to be on the safe side, before transmitting the postal code) is a better user experience. The way I imagine it is that on blur there's a high chance the user will notice the automatic formatting - and it will not be a surprise to them later (in case someone is used to using `w1k1qz` and is surprised by seeing `W1K 1QZ` later on the receipt or something). Since they will see the reformatting in the context of the field. And worst case scenario, will try to reformat again manually - only to get it re-reformatted. And hopefully understand that we simply want that format.

I'd pull in @fditrapani here for a gut check 🙏

**Comment by sirbrillig** [comment]
Sounds good. Added reformatting on blur explicitly this time.

**Comment by sirbrillig** [comment]
(Failing tests are unrelated.)

**Pull request merged**